pydantic/fields.py

```diff
@@ -895,6 +947,36 @@ def _validate_singleton(
     ) -> 'ValidateReturn':
         if self.sub_fields:
             errors = []
+
+            if get_origin(self.type_) is Union and self.discriminated_union_config is not None:
+                if not self.discriminated_union_config.sub_fields_mapping:
+                    assert cls is not None
+                    raise ConfigError(
+                        f'field "{self.name}" not yet prepared so type is still a ForwardRef, '
+                        f'you might need to call {cls.__name__}.update_forward_refs().'
+                    )
+
+                discriminator_key = self.discriminated_union_config.discriminator_key
+
+                try:
+                    discriminator_value = v[discriminator_key]
```

Comment:
Not sure this is intentional - this assumes `v` is a dict, if `v` is a class inherited from `pydantic.BaseModel`, it will fail:

```python
>>> import pydantic, typing
...
... class A(pydantic.BaseModel):
...     l: typing.Literal["a"]
...
... class B(pydantic.BaseModel):
...     l: typing.Literal["b"]
...
... class Top(pydantic.BaseModel):
...     sub: typing.Union[A, B] = pydantic.Field(..., discriminator="l")
...
... t = Top(sub=A(l="a"))
Traceback (most recent call last):
  File "<stdin>", line 12, in <module>
  File "/Users/krisb/Library/Caches/pypoetry/virtualenvs/project-py3.8/lib/python3.8/site-packages/pydantic/main.py", line 400, in __init__
    raise validation_error
pydantic.error_wrappers.ValidationError: 1 validation error for Top
sub
  Discriminator 'l' is missing in value (type=value_error)

1 validation error for Top
sub
  Discriminator 'l' is missing in value (type=value_error)
>>> t = Top(**{"sub": {"l": "a"}})
>>> t
Top(sub=A(l='a'))

>>> class TopWithoutDiscriminator(pydantic.BaseModel):
...     sub: typing.Union[A, B]
... t = TopWithoutDiscriminator(sub=A(l="a"))
>>> t
TopWithoutDiscriminator(sub=A(l='a'))
```

Reply:
Thanks a lot @kbakk! It was definitely not intentional ;)